File: yaf-workflow-controller/src/main/java/ar/edu/ungs/yamiko/workflow/serial/SerialGA.java
package ar.edu.ungs.yamiko.workflow.serial;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.Iterator;
import java.util.List;
import java.util.Set;

import org.apache.log4j.Logger;

import ar.edu.ungs.yamiko.ga.domain.Individual;
import ar.edu.ungs.yamiko.ga.domain.Population;
import ar.edu.ungs.yamiko.ga.exceptions.InvalidProbability;
import ar.edu.ungs.yamiko.ga.exceptions.NullAcceptEvaluator;
import ar.edu.ungs.yamiko.ga.exceptions.NullCrossover;
import ar.edu.ungs.yamiko.ga.exceptions.NullFitnessEvaluator;
import ar.edu.ungs.yamiko.ga.exceptions.NullPopulationInitializer;
import ar.edu.ungs.yamiko.ga.exceptions.NullSelector;
import ar.edu.ungs.yamiko.ga.exceptions.YamikoException;
import ar.edu.ungs.yamiko.ga.toolkit.StaticHelper;
import ar.edu.ungs.yamiko.workflow.BestIndHolder;
import ar.edu.ungs.yamiko.workflow.Parameter;

public class SerialGA<T> {

		private Parameter<T> parameter;
		private long generationNumber=0;
		private double bestFitness=0;
		private Individual<T> bestInd;
		private Population<T> finalPopulation;

		
		
		public Population<T> getFinalPopulation() {
			return finalPopulation;
		}

		public SerialGA(Parameter<T> parameter) throws YamikoException{
			super();
			this.parameter = parameter;
			validateParameters();
		}
		
		private void validateParameters()
		{
			if (parameter.getAcceptEvaluator()==null) throw new NullAcceptEvaluator();
			if (parameter.getCrossover()==null) throw new NullCrossover() ;
			if (parameter.getCrossoverProbability()<=0 || parameter.getCrossoverProbability()>1) throw new InvalidProbability() ;
			if (parameter.getMutationProbability()<=0 || parameter.getMutationProbability()>1) throw new InvalidProbability() ;
			if (parameter.getFitnessEvaluator()==null) throw new NullFitnessEvaluator() ;
			if (parameter.getPopulationInitializer()==null) throw new NullPopulationInitializer() ;
			if (parameter.getSelector()==null) throw new NullSelector() ;
		}
		
		@SuppressWarnings({ "rawtypes", "unchecked" })
		public Individual<T> run() throws YamikoException
		{
			Population<T> p=parameter.getPopulationInstance();
			parameter.getPopulationInitializer().execute(p);
			
			
			while (generationNumber<parameter.getMaxGenerations() && parameter.getOptimalFitness()>bestFitness)
			{
				Individual<T> bestIndGen=p.iterator().next();
				double fitnessProm=0d;
				int cont=0;
				for (Individual<T> individual : p)
				{
					if (individual.getFitness()==null)
					{
						parameter.getMorphogenesisAgent().develop(parameter.getGenome(),individual);
						individual.setFitness(parameter.getFitnessEvaluator().execute(individual));
					}
					if (individual.getFitness()>bestFitness)
					{
						bestFitness=individual.getFitness();
						bestInd=individual;
						BestIndHolder.holdBestInd(bestIndGen);						
					}
					if (individual.getFitness()>bestIndGen.getFitness())
					{
						bestIndGen=individual;
						BestIndHolder.holdBestInd(bestIndGen);						
					}
					fitnessProm+=individual.getFitness();
					cont++;
				}
				fitnessProm=fitnessProm/cont;
				//if ((generationNumber % 100)==0) 
					Logger.getLogger("file").warn("Mejor Individuo Generación " + generationNumber + " Fitness: " + bestIndGen.getFitness() + " -- fitness promedio: " + fitnessProm + "{[" + bestIndGen.getFitness() + ";" + fitnessProm +"]}");

				parameter.getSelector().setPopulation(p);
				List<Individual<T>> candidates=parameter.getSelector().executeN(p.size().intValue()*2);
				
				Iterator ite=candidates.iterator();
				Set<Individual<T>> replacedChildren=new HashSet<Individual<T>>();
				replacedChildren.add(bestIndGen);
				while (ite.hasNext()) {
					Individual<T> parentA=(Individual<T>)ite.next();
					if (StaticHelper.randomDouble(1d)<=parameter.getCrossoverProbability() && ite.hasNext())
					{
						Individual<T> parentB=(Individual<T>)ite.next();
						List<Individual<T>> parents=new ArrayList<Individual<T>>();
						
						parents.add(parentA);
						parents.add(parentB);
						List<Individual<T>> children=parameter.getCrossover().execute(parents);
						List<Individual<T>> acceptedChildren=parameter.getAcceptEvaluator().execute(children, parents);
						for (int i=0;i<acceptedChildren.size();i++)
							replacedChildren.add(acceptedChildren.get(i));
					}		
				}
				p.replacePopulation(replacedChildren);
						
				for (Individual<T> ind : p) 
					if (ind.getId()!=bestIndGen.getId())
						if (StaticHelper.randomDouble(1d)<=parameter.getMutationProbability())
							parameter.getMutator().execute(ind);
				
				generationNumber++;
				
//				if ((generationNumber % 100)==0) 
//					Logger.getLogger("file").warn("Generation " + generationNumber + "...");
				
				
			}
			
			Logger.getLogger("file").warn("... Cumplidas " + generationNumber + " Generaciones.");

			for (Individual<T> individual : p)
				if (individual.getFitness()==null)
				{
					parameter.getMorphogenesisAgent().develop(parameter.getGenome(),individual);
					individual.setFitness(parameter.getFitnessEvaluator().execute(individual));
				}
			finalPopulation=p;

			return bestInd;
			
		}
		
		
	
}

File: yaf-workflow-controller-scala/src/main/scala/ar/edu/ungs/yamiko/workflow/serial/SerialGA.scala
package ar.edu.ungs.yamiko.workflow.serial

import org.apache.log4j.Logger
import ar.edu.ungs.yamiko.ga.domain.Genome
import ar.edu.ungs.yamiko.ga.domain.Individual
import ar.edu.ungs.yamiko.ga.operators.AcceptEvaluator
import ar.edu.ungs.yamiko.ga.operators.Crossover
import ar.edu.ungs.yamiko.ga.operators.FitnessEvaluator
import ar.edu.ungs.yamiko.ga.operators.MorphogenesisAgent
import ar.edu.ungs.yamiko.ga.operators.Mutator
import ar.edu.ungs.yamiko.workflow.BestIndHolder
import ar.edu.ungs.yamiko.workflow.Parameter
import scala.collection.mutable.ListBuffer
import ar.edu.ungs.yamiko.ga.operators.PopulationInitializer
import ar.edu.ungs.yamiko.ga.domain.impl.DistributedPopulation
import ar.edu.ungs.yamiko.ga.exceptions.YamikoException
import scala.util.Random
import java.text.DecimalFormat
import ar.edu.ungs.yamiko.ga.exceptions.YamikoException
import ar.edu.ungs.yamiko.workflow.parallel.spark.scala.ParameterValidator
import scala.collection.TraversableOnce.flattenTraversableOnce



class SerialGA[T] (parameter: Parameter[T]) extends Serializable{
  
  private var _finalPop:List[Individual[T]] = null
  def finalPopulation:List[Individual[T]] = _finalPop 
  private val r:Random=new Random(System.currentTimeMillis()) 
  private var bestIndHolder=new BestIndHolder[T]()
  private val notScientificFormatter:DecimalFormat = new DecimalFormat("#");
  def getBestIndHolder()=bestIndHolder
  
  @throws(classOf[YamikoException])
  def run():Individual[T] =
		{
      ParameterValidator.validateParameters(parameter);
    	var generationNumber=0;
		  var bestFitness:Double=0;
		  var bestInd:Individual[T]=null;

			val startTime=System.currentTimeMillis()
			
			var population=parameter.getPopulationInstance();
			
			// Si no vino inicializada la poblacion, se la inicializa
			if (population.getAll().size==0) parameter.getPopulationInitializer().execute(population)
			
			while (generationNumber<parameter.getMaxGenerations() && parameter.getOptimalFitness()>bestFitness)
			{
			  generationNumber+=1
			  if (generationNumber%10==0) Logger.getLogger("file").warn("Generation " + generationNumber + " -> principio del bucle");
			  
        val t1=System.currentTimeMillis()

        val descendants=new ListBuffer[Individual[T]]
		    population.getAll().foreach { i:Individual[T] => 
  			        if (i.getFitness()==0)
  				      {
                    if (i.getPhenotype==null)  parameter.getMorphogenesisAgent().develop(parameter.getGenome(),i)
  					        i.setFitness(parameter.getFitnessEvaluator().execute(i))
  					    }
		        
        //if (g%10==0) Logger.getLogger("file").warn("Generation población " + dp.getId() + " - " +g + " -> developed");
        val bestOfGeneration=population.getAll().maxBy { x => x.getFitness }    			        
				bestIndHolder.holdBestInd(bestOfGeneration);				
				if (bestOfGeneration.getFitness()>bestFitness)
				{
					bestFitness=bestOfGeneration.getFitness();
					bestInd=bestOfGeneration;					
				}

        // Profiler
        if (generationNumber%10==0) Logger.getLogger("profiler").debug(generationNumber+";"+bestOfGeneration.getId()+";"+notScientificFormatter.format(bestOfGeneration.getFitness())+";"+System.currentTimeMillis())

	      val candidates:List[Individual[T]]=(parameter.getSelector().executeN((population.size()).intValue(),population)).asInstanceOf[List[Individual[T]]];
				val tuplasSer=candidates.sliding(1, 2).flatten.toList zip candidates.drop(1).sliding(1, 2).flatten.toList

    		for (t <- tuplasSer)
    				{
		            if (t._1.getPhenotype==null) parameter.getMorphogenesisAgent().develop(parameter.getGenome(), t._1 )
		            if (t._1.getFitness==0) t._1.setFitness(parameter.getFitnessEvaluator().execute(t._1))
		            if (t._2.getPhenotype==null) parameter.getMorphogenesisAgent().develop(parameter.getGenome(), t._2 )
		            if (t._2.getFitness==0) t._2.setFitness(parameter.getFitnessEvaluator().execute(t._2))
		            val parentsJ=List(t._1,t._2)
      				  val desc=parameter.getCrossover().execute(parentsJ)
      				  for (d <- desc)
      				  {
			            if (d.getPhenotype==null) parameter.getMorphogenesisAgent().develop(parameter.getGenome(), d )
			            if (d.getFitness==0) d.setFitness(parameter.getFitnessEvaluator().execute(d))
      				  }
      				  for (d <- parameter.getAcceptEvaluator().execute(desc,parentsJ))
      				  {
      				    if (r.nextDouble()<=parameter.getMutationProbability()) parameter.getMutator().execute(d)
			            if (d.getPhenotype==null) parameter.getMorphogenesisAgent().develop(parameter.getGenome(), d )
			            if (d.getFitness==0) d.setFitness(parameter.getFitnessEvaluator.execute(d))
			            descendants+=d
    					  }      				    
    				}
    				if (generationNumber==1)
    				if (!descendants.contains(bestOfGeneration))
    				{
		          //Logger.getLogger("file").warn("Generation población " + dp.getId() + " - " +g + " -> No contenía al mejor de la generación " + bestOfGeneration.getId + " - " + bestOfGeneration.getFitness);
    				  descendants.dropRight(1)
    				  descendants+=(bestOfGeneration)
    				}

			  			  
			  // Ordenar por fitness
			  population.replacePopulation(population.getAll().sortBy(_.getFitness).reverse)}
			  if (generationNumber%100==0)
			  {
			    Logger.getLogger("file").warn("Generación " + generationNumber + " - Finalizada - Transcurridos " + (System.currentTimeMillis()-startTime)/1000d + "'' - 1 Generación cada " + (System.currentTimeMillis().doubleValue()-startTime.doubleValue())/generationNumber  + "ms"  )
			    println("Generación " + generationNumber + " - Mejor Elemento total " + bestInd.getFitness)
			  }
			}

			Logger.getLogger("file").info("... Cumplidas " + generationNumber + " Generaciones.");

      _finalPop=population.getAll()

      return bestInd;

		}
}